controller/update.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	if numRunning < numTotal+opts.MaxGrowth {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (c controller) addWorker(jobDesc updateJobDesc, removeQueue chan updateJobDesc, fail chan<- error) {
+	// Submit.
+	taskObject, err := c.Submit(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer := make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+
+	// Start.
+	taskObject, err = c.Start(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer = make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+
+	removeQueue <- jobDesc
+}
+
+func (c controller) removeWorker(jobDesc updateJobDesc, fail chan<- error) {
+	// Stop.
+	taskObject, err := c.Stop(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer := make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+
+	// Destroy.
+	taskObject, err = c.Destroy(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer = make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+}
+
+// UpdateOptions represents the options defining the strategy of an update
+// process. Lets have a look at how the update process of 3 group slices would
+// look like using the given options.
+//
+//     TODO I am not that happy with this visualization. Improving it? Removing it?
+//
+//     MaxGrowth    1
+//     MinAlive     2
+//     ReadySecs    30
+//
+//     @1 (running)  ->  @1 (stopped/destroyed)
+//     @2 (running)  ->  @2 (running)            ->  @2 (stopped/destroyed)
+//     @3 (running)  ->  @3 (running)            ->  @3 (running)            ->  @3 (stopped/destroyed)
+//                   ->  @1 (submitted/running)  ->  @1 (running)            ->  @1 (running)            ->  @1 (running)
+//                                               ->  @2 (submitted/running)  ->  @2 (running)            ->  @2 (running)
+//                                                                           ->  @3 (submitted/running)  ->  @3 (running)
+//
+type UpdateOptions struct {
+	// MaxGrowth represents the number of groups allowed to be added at a given
+	// time. No more than MaxGrowth groups will be added at the same time during
+	// the update process.
+	MaxGrowth int
```

Comment:
which reminds me: We never actually discussed if we want this possibility of kubernetes to also specify percentage values for this. I guess if we want this, we can do it in a later PR. We should make an issue for this.